### ISSUE TITLE
Restrict Navigation Outside Initial Directory in mini.files

### ISSUE THREAD

**Issue by ahmedreda1009** [echasnovski/mini.nvim#1162]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

mini.files

### Description

I am using the mini.files plugin and would like to restrict navigation such that when using the go_out() function, I want to prevent navigation outside the initial directory from which mini.files was opened. I have attempted to implement it myself, but it seems to not work as expected.

In other words I would like the plugin to restrict navigation so that when I try to go up from the initial directory, it prevents me from going outside this directory.

**Comment by echasnovski** [comment]
Thanks for the suggestion!

I think the best approach here would be via custom mapping which would essentially:
- Do nothing is currently focused path is the initial (anchor) directory.
- Execute `go_out()` otherwise.

The main issue here is programmatically determining what is the the anchor directory for currently opened explorer. It *can* mostly be done inside 'MiniFilesExplorerOpen' event (via calling `get_fs_entry()` and computing its directory name), but it is a bit cumbersome and not 100% reliable. I'll think about making it more straightforward.

**Comment by ahmedreda1009** [comment]
Thank you for the guidance on how to approach the feature! I was able to implement a custom solution that restricts navigation within the initial directory. Here’s the code I used:


```
-- Prevent the go_out function from going outside the entry directory
local anchor_dir = ""
local is_mini_files_explorer_open_event_triggered = false

-- Get the entry directory path
vim.api.nvim_create_autocmd("User", {
  pattern = "MiniFilesExplorerOpen",
  callback = function()
    if not is_mini_files_explorer_open_event_triggered then
      local anchor_path = MiniFiles.get_fs_entry().path
      anchor_dir = vim.fs.dirname(anchor_path)
      is_mini_files_explorer_open_event_triggered = true
    end
  end
})

-- Custom function to override go_out()
local function custom_go_out()
  local current_path = MiniFiles.get_fs_entry().path
  local current_dir = vim.fs.dirname(current_path)
  if current_dir == anchor_dir then
    print("Cannot navigate outside the initial directory.")
  else
    require("mini.files").go_out()
  end
end

-- Map the custom go_out function
vim.keymap.set("n", "h", custom_go_out, { noremap = true, silent = true })

-- don't forget to unmap the h key before remapping:
    require('mini.files').setup({
      mappings = {
        go_out = '',
      },
    })

```

This approach works well for my use case, and I hope it can be helpful to others who might want similar functionality. Thanks again for the help!

**Comment by echasnovski** [comment]
> vim.keymap.set("n", "h", custom_go_out, { noremap = true, silent = true })

Making a global mapping like that is probably not a good idea. I'd recommend adapting [this example](https://github.com/echasnovski/mini.nvim/blob/a8185957febe7dae31556dbd7326cfa597b812f1/doc/mini-files.txt#L450-L457) and create a buffer-local mapping inside autocommand for `MiniFilesBufferCreate` event. This would mean moving all the helper logic inside `MiniFilesExplorerOpen` autocommand.

I do plan to make this use case a bit more straightforward. Probably, by introducing `get_explorer_data()` function which can be used to get anchor path without the need of 'MiniFilesExplorerOpen' workaround.

**Comment by ahmedreda1009** [comment]
Thank you so much for your help and guidance on this issue! I apologize for deleting my earlier post before seeing your reply—I was troubleshooting the mapping issues and ended up removing it.

I’m happy to report that I’ve now finalized the solution, and it’s working perfectly. Here’s the code I’m using to prevent `mini.files` from navigating outside the initial directory:

```lua
-- Prevent `go_out` from going outside the entry directory
local anchor_dir = ""
local is_mini_files_explorer_open_event_triggered = false

local function get_curr_dir()
  local curr_path = MiniFiles.get_fs_entry().path
  local curr_dir = vim.fs.dirname(curr_path)
  return curr_dir
end

-- Get the entry directory path per session
vim.api.nvim_create_autocmd("User", {
  pattern = "MiniFilesExplorerOpen",
  callback = function()
    if not is_mini_files_explorer_open_event_triggered then
      anchor_dir = get_curr_dir()
      is_mini_files_explorer_open_event_triggered = true
    end
  end
})

-- Custom function to override `go_out`
local function custom_go_out()
  local current_dir = get_curr_dir()
  if current_dir == anchor_dir then
    print("Cannot navigate outside the initial directory.")
  else
    require("mini.files").go_out()
  end
end

-- Map the custom `go_out` function
vim.api.nvim_create_autocmd('User', {
  pattern = 'MiniFilesBufferCreate',
  callback = function(args)
    local buf_id = args.data.buf_id
    vim.keymap.set('n', 'h', custom_go_out, { buffer = buf_id })
  end,
})
```

This approach ensures that the `h` mapping is overridden per buffer and works reliably. I hope this can serve as a reference for anyone looking to implement similar functionality before your more sophisticated solution is available.

Thanks again for your support!

**Comment by echasnovski** [comment]
@ahmedreda1009, on latest `main` there is now [`get_explorer_state()`](https://github.com/echasnovski/mini.nvim/blob/73f2e7a596bbb4e6ae6728d3a4426221ffc9512d/doc/mini-files.txt#L807) function. It can be used inside custom buffer-local mappings to have a more straightforward "don't go outside of initial path" logic. For example, with something like this:

```lua
require('mini.files').setup({ mappings = { go_out = '' } })

-- Make `go_out` to not go outside of initial path (a.k.a. "anchor")
local go_out_restricted = function()
  local state = MiniFiles.get_explorer_state()
  local cur_win_id, cur_path = vim.api.nvim_get_current_win(), nil
  for _, win_data in ipairs(state.windows) do
    if win_data.win_id == cur_win_id then cur_path = win_data.path end
  end
  if cur_path == state.anchor then return print('Cannot navigate outside the initial directory.') end
  MiniFiles.go_out()
end

local map_go_out = function(args)
  local rhs = function()
    for _ = 1, vim.v.count1 do
      go_out_restricted()
    end
  end
  vim.keymap.set('n', 'h', rhs, { buffer = args.data.buf_id })
end
vim.api.nvim_create_autocmd('User', { pattern = 'MiniFilesBufferCreate', callback = map_go_out })
```

**Comment by ahmedreda1009** [comment]
Hi @echasnovski ,

Thank you so much for adding the get_explorer_state() function and providing the example for the “don’t go outside of initial path” logic. This is exactly what I needed, and the implementation looks great! I really appreciate your quick response and the continuous improvements you’re making to mini.files. It’s been a game-changer for my workflow.

Thanks again for your help and support!
